kythe/rust/indexer/src/indexer/analyzers.rs

```diff
@@ -0,0 +1,270 @@
+// Copyright 2020 The Kythe Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::error::KytheError;
+use crate::writer::KytheWriter;
+
+use super::entries::EntryEmitter;
+
+use analysis_rust_proto::CompilationUnit;
+use rls_analysis::Crate;
+use rls_data::{Def, DefKind};
+use std::collections::HashMap;
+use std::fs::File;
+use std::io::Read;
+use std::path::{Path, PathBuf};
+use storage_rust_proto::*;
+
+/// A data structure to analyze and index CompilationUnit protobufs
+pub struct UnitAnalyzer<'a> {
+    unit: &'a CompilationUnit,
+    unit_storage_vname: VName,
+    emitter: EntryEmitter<'a>,
+    root_dir: &'a PathBuf,
+    file_vnames: HashMap<String, VName>,
+}
+
+/// A data structure to analyze and index individual crates
+pub struct CrateAnalyzer<'a, 'b> {
+    emitter: &'b mut EntryEmitter<'a>,
+    file_vnames: &'b HashMap<String, VName>,
+    unit_vname: &'b VName,
+    krate: Crate,
+    krate_ids: HashMap<u32, String>,
+}
+
+impl<'a> UnitAnalyzer<'a> {
+    /// Create an instance to assist in analyzing `unit`. Graph information will
+    /// be written to the `writer` and source file contents will be read using
+    /// `root_dir` as a base directory.
+    pub fn new(
+        unit: &'a CompilationUnit,
+        writer: &'a mut dyn KytheWriter,
+        root_dir: &'a PathBuf,
+    ) -> Self {
+        // Create a HashMap between the file path and the VName which we can retrieve
+        // later to emit nodes
+        let mut file_vnames = HashMap::new();
+        for required_input in unit.get_required_input() {
+            let analysis_vname = required_input.get_v_name();
+            // The required input has a VName of type analysis_rust_proto::analysis::VName.
+            // We need storage_rust_proto::storage::VName to create an Entry. Therefore, we
+            // must use unsafe rust to transmute the VName.
+            let storage_vname: VName = unsafe { ::std::mem::transmute(analysis_vname.clone()) };
```

Comment:
Please pull out the VName message conversions to separate functions and if possible get rid of uses of `unsafe` (e.g., by constructing VName instances by manually copying the five fields).